plugin: Introduce protocol and terraform-plugin-mux documentation (#2181)
hashicorp/terraform-website: content/plugin/mux/translating-protocol-version-5-to-6.mdx

---
page_title: 'Plugin Development - Translating Protocol Version 5 to 6'
description: >-
  Use the tf5to6server package in terraform-plugin-mux to translate protocol version 5 providers to protocol version 6.
---

# Translating Protocol Version 5 to 6

The [`tf5to6server`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-mux/tf5to6server) package enables translating a [protocol version 5 provider server](../how-terraform-works#protocol-version-5) into a [protocol version 6 provider server](../how-terraform-works#protocol-version-6).

Use this package to:

> * Migrate individual resources and data sources from [terraform-plugin-sdk/v2](../sdkv2/sdk-intro) to [terraform-plugin-framework](../framework) over time, while taking advantage of [protocol version 6](../how-terraform-works#protocol-version-6) features in terraform-plugin-framework.

Reviewer: We just linked to the protocol 6 stuff, so we probably don't need it again here so soon. 

```suggestion
* Migrate individual resources and data sources from [terraform-plugin-sdk/v2](../sdkv2/sdk-intro) to [terraform-plugin-framework](../framework) over time, while using protocol version 6 features in terraform-plugin-framework.
```

* Develop with [terraform-plugin-sdk/v2](../sdkv2/sdk-intro), but require Terraform CLI 1.0 or later.

## Compatibility

Protocol version 6 provider servers are compatible with Terraform CLI 1.0 and later.

~> **NOTE**: If upgrading [terraform-plugin-sdk/v2](../sdkv2/sdkv2-intro), Terraform CLI 1.1.5 and later is required.

The following provider server implementations are supported:

* [terraform-plugin-sdk/v2](../sdkv2/sdkv2-intro)
* [terraform-plugin-go tf5server](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server)

## Requirements

Upgrading provider servers from protocol version 5 to protocol version 6 has no provider code requirements.

> If publishing to the [Terraform Registry](/registry), the [Terraform Registry manifest file](https://www.terraform.io/registry/providers/publishing#terraform-registry-manifest-file) should have `metadata.protocol_versions` set to `["6.0"]`.

Reviewer: ```suggestion
If publishing to the [Terraform Registry](/registry), set `metadata.protocol_versions` to `["6.0"]` in the [Terraform Registry manifest file](https://www.terraform.io/registry/providers/publishing#terraform-registry-manifest-file).
```

## Code Implementation

> Use the [`tf5to6server.UpgradeServer()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-mux/tf5to6server#UpgradeServer) function to wrap a provider server. For most providers, this is in the provider `main()` function of the root directory `main.go` file or where [`tf6muxserver`](./combining-protocol-version-6-providers) is implemented in the codebase, if applicable.

Reviewer: ```suggestion
Use the [`tf5to6server.UpgradeServer()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-mux/tf5to6server#UpgradeServer) function to wrap a provider server. For most providers, this is either in the provider `main()` function of the root directory `main.go` file or where [`tf6muxserver`](./combining-protocol-version-6-providers) is implemented in the codebase.
```
